### PR TITLE
Fix bugs where unique_range became invalid

### PR DESCRIPTION
**Pull request by saethlin**
And also expand the cache integrity checks to cover this case.

I'm going to run this over all the ICEs I've gotten out of Miri recently, could be a bit.

Fixes https://github.com/rust-lang/miri/issues/2389

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks!

Is there a self-contained testcase we could add as regression test?

**Comment by saethlin** [comment]
Actually I only had 23 examples, it looks like a previous time I ran this was with a rustc version that had a very easy-to-hit ICE issue in `format!`.

I think the bad pattern is
1. Put `Unique` items in the stack
2. Call `set_unknown_bottom`
3. Do a read, which tries to skip past everything before `unique_range`, which we should have zeroed in step 2
4. Invalid range is passed to slice indexing, panic

**Comment by RalfJung** [comment]
I was hoping there'd be some example in your pool that is reasonably easy to minimize.

I tried triggering that pattern but so far it doesn't ICE.

**Comment by RalfJung** [comment]
What is the point of this `min` here anyway?
```rust
            let upper = (unique_range.end + 1).min(self.borrows.len());
```
If the range is in-bounds it shouldn't need capping. Also why the `+1`?

**Comment by RalfJung** [comment]
Is it even possible for `unique_range.start` to be anything but 0? It is initially 0, and would grow only if `unique_range.start >= new_idx`, but we don't usually insert at index 0...

**Comment by RalfJung** [comment]
Maybe the initial unique_range for allocations with a non-unique initial item should be `1..1`, that seems to me like it'd lead to more precise tracking.

**Comment by RalfJung** [comment]
Aha, got a reproducer:
```rust
use std::cell::Cell;

fn main() { unsafe {
    let root0 = Cell::new(42);
    let raw = (&root0 as *const Cell<i32> as usize as *const Cell<i32>);
    // empty the stack to unknown (via SRW reborrow from wildcard)
    let ref0 = &*raw;
    // Do a non-SRW reborrow from wildcard to start building up a stack again.
    // Now new refs start being inserted at idx 0, pushing the unique_range up.
    let refN = &*&*&*&*&*(raw.cast::<i32>());
    // empty the stack again, but this time with unique_range sitting at some high index.
    let ref0 = &*raw;
    // and do a read which tries to clear the uniques
    raw.cast::<i32>().read();
} }
```
That took a while... and "Put Unique items in the stack" lead me down the totally wrong track. It doesn't seem enough to get just the unique_range.end beyond the stacksize, I need to also get the beginning beyond the stacksize, and growing unique_range.start is actually not easy (hence my previous question).

**Comment by saethlin** [comment]
Just to be clear, I think it is worthwhile to have a regression test for this, but simply adding the extra check inside `verify_cache_consistency` causes a few tests to fail. So the existing test suite does produce inconsistent state with the cache, it just wasn't detected by anything. I don't know if the fact that the tests would have exposed this with a better consistency check was already clear.

---

> If the range is in-bounds it shouldn't need capping. Also why the `+1`?

Well the `+1` exists because at one point in the middle of implementing this the range was inclusive. It's not in the existing code, so yeah that should definitely not have made it into the implementation.

I suspect it was also capped because in the past I've thought of this as "just an optimization" so it would be okay for it to be a bit sloppy, as you alluded to in the issue

> If it was just unnecessarily big but still in-bounds, that would be fine (just inefficient).

But I think it's clear based on the current situation that line of thinking wasn't very wise. I'm trying to tighten up the consistency checks a whole lot more and remove the slop from the implementation.

**Comment by RalfJung** [comment]
> But I think it's clear based on the current situation that line of thinking wasn't very wise. I'm trying to tighten up the consistency checks a whole lot more and remove the slop from the implementation.

I'd prefer if you left that for a future PR. I hope to get this one landed sooner rather than later and then bump the Miri submodule in time for the next nightly in ~23h.

This one should make the consistency check correct (i.e., ensure the unique_range is in-bounds), and add the ICE test, and maybe clean up the computation of `upper` that I brought up. And ensure there are no other leftovers from "unique_range is inclusive".

**Comment by RalfJung** [comment]
Oh, you already did it. Oh well.^^

**Comment by saethlin** [comment]
I can edit the changes out, if you want. It's not a whole lot.

File: src/stacked_borrows/stack.rs [hunk 1]
```diff
@@ -102,6 +103,18 @@ impl<'tcx> Stack {
...
+        // Check that the unique_range is a valid index into the borrow stack.
+        let uniques = &self.borrows[self.unique_range.clone()];
+
+        // Check that the start of the unique_range is precise.
+        if let Some(first_unique) = uniques.first() {
+            assert_eq!(first_unique.perm(), Permission::Unique);
+        }
+        // We cannot assert that the unique range is exact on the upper end.
+        // When we pop items within the unique range, setting the end of the range precisely
+        // require doing a linear search of the borrow stack, which is exactly the kind of
```
Comment (RalfJung) [inline src/stacked_borrows/stack.rs:116]:
```suggestion
        // requires doing a linear search of the borrow stack, which is exactly the kind of
```

File: src/stacked_borrows/stack.rs [hunk 2]
```diff
@@ -227,9 +240,14 @@ impl<'tcx> Stack {
             self.unique_range.end += 1;
         }
         if new.perm() == Permission::Unique {
-            // Make sure the possibly-unique range contains the new borrow
-            self.unique_range.start = self.unique_range.start.min(new_idx);
-            self.unique_range.end = self.unique_range.end.max(new_idx + 1);
+            // If this is the first Unique, set the range to contain just the new item.
+            if self.unique_range == (0..0) {
```
Comment (RalfJung) [inline src/stacked_borrows/stack.rs:244]: Can this be generalized to "if the range is empty"?

File: tests/pass/issues/issue-miri-2389.rs [hunk 3]
```diff
@@ -0,0 +1,17 @@
+use std::cell::Cell;
+
+fn main() {
```
Comment (RalfJung) [inline tests/pass/issues/issue-miri-2389.rs:3]: Can you move this into the Stacked Borrows folder?

File: src/stacked_borrows/stack.rs [hunk 4]
```diff
@@ -102,6 +103,18 @@ impl<'tcx> Stack {
                 );
             }
         }
+
+        // Check that the unique_range is a valid index into the borrow stack.
+        let uniques = &self.borrows[self.unique_range.clone()];
```
Comment (RalfJung) [inline src/stacked_borrows/stack.rs:108]: This will also assert `start <= end` I assume?

**Comment by RalfJung** [comment]
Great, thanks a lot. :)
@bors r+

**Pull request merged**